app/Http/Controllers/v1/User/ContributionController.php

```diff
@@ -0,0 +1,171 @@
+	        ['only' => ['create']]
+        );
+        $this->middleware(
+            \Scopes::matchOneOfDeepestChildren('user-edit-contributions', 'client-edit-contributions'),
+	        ['only' => ['edit']]
+        );
+        $this->middleware(
+            \Scopes::matchOneOfDeepestChildren('user-remove-contributions', 'client-remove-contributions'),
+        	['only' => ['remove']]
+        );
+    }
+
+    /**
+     * List contributions of the current user.
+     *
+     * @param Request $request
+     * @param string  $user_id
+     * @return JsonResponse
+     */
+    public function index(Request $request, string $user_id=null)
+    {
+        $user = $this->getUser($request, $user_id);
+        $ginger = \Ginger::userByEmail($user->email);
+        $contributions = [];
+        $semesters = Semester::get()
+            ->map(function($semester){
+                return $semester->hideData();
+            })->toArray();
+
+        foreach ($ginger->getContributions() as $contribution) {
+            $contributionSemesters = array_filter($semesters, function ($semester) use ($contribution) {
+                if (($semester['begin_at'] < $contribution->end_at) && ($semester['end_at'] > $contribution->begin_at)) {
+                    return true;
+                }
+            });
+
+            $newContribution = [
+                'start' => $contribution->begin_at,
+                'end' => $contribution->end_at,
+                'amount' => $contribution->money,
+                'semesters' => array_values($contributionSemesters),
+
+            ];
+
+            $contributions[] = $newContribution;
+        }
+
+        return response()->json($contributions);
+    }
+
+    /**
+     * Show a contribution depending on a given semester range and an optionnel user.
+     *
+     * @param Request $request
+     * @param string  $semester_id Work with id or name.
+     * @param string  $user_id
+     * @return JsonResponse
+     */
+    public function show(Request $request, string $semester_id="current", string $user_id=null)
+    {
+        $semester = $this->getSemester($semester_id, true);
+        $user = $this->getUser($request, $user_id);
+        $contributions = \Ginger::userByEmail($user->email)->getContributions();
+
+        $contribution = Arr::first($contributions, function ($contribution) use ($semester){
+            $dateToContain = ((strtotime($semester->begin_at) + strtotime($semester->end_at)) / 2);
+            $contributionStart = strtotime($contribution->begin_at);
+            $contributionEnd = strtotime($contribution->end_at);
+
+            if ($contributionStart < $dateToContain && $dateToContain < $contributionEnd) {
+                return true;
+            }
+        });
+
+        if (!is_null($contribution)) {
+            $semesters[] = $semester;
+            if ($semester->is_spring) {
+                array_unshift($semesters, $this->getSemester(
```

Comment:
Wow ok i see what you did here, I found that a bit weird.

I think the best would be to get in function of dates.
But a better way here to avoid the horrible: `strval(intval(substr($semester->name, -2)) - 1)` is that the semester contains the year.
So you can do: 
```php
$this->getSemester('A|P'.((string) $semester->year -|+ 1);
```

Reply:
Avoid absotly those `strval` and `intval`, I found that disturbing. Prefer auto conversion and dynamic typing like: `(string) 42`